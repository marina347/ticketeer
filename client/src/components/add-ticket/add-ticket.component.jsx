import React, { useState } from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";

import FormTextArea from "../form-inputs/form-text-area/form-text-area.component";
import FormButton from "../form-button/form-button.component";
import { addTicketAsync } from "../../redux/ticket/ticket.actions";
import {
  selectToken,
  selectCurrentUser,
} from "../../redux/user/user.selectors";
import "./add-ticket.styles.scss";

export const AddTicket = ({ addTicket, _id, token }) => {
  const [ticketName, setTicketName] = useState("");

  const handleChange = (event) => {
    const { value } = event.target;
    setTicketName(value);
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    addTicket(
      {
        name: ticketName,
        laneId: _id,
      },
      token
    );
    setTicketName("");
  };

  return (
    <form id="add_ticket_form" className="add-ticket" onSubmit={handleSubmit}>
      <FormTextArea
        id="add_ticket_input"
        name="ticketName"
        label="Ticket name"
        placeholder="Ticket name"
        cols={50}
        rows={3}
        value={ticketName}
        onChange={handleChange}
      />
      <FormButton
        className="btn btn-main btn--small u-animation-none"
        type="submit"
      >
        ADD
      </FormButton>
    </form>
  );
};

const mapStateToProps = createStructuredSelector({
  token: selectToken,
  currentUser: selectCurrentUser,
});

const mapDispatchToProps = (dispatch) => ({
  addTicket: (ticket, token) => dispatch(addTicketAsync(ticket, token)),
});

export default connect(mapStateToProps, mapDispatchToProps)(AddTicket);
