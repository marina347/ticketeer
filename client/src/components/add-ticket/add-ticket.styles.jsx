import styled from "styled-components";
import { AddLaneContainer } from "../add-lane/add-lane.styles";

export const AddTicketItemContainer = styled.div`
  text-align: center;
  margin: 20px;
`;

AddTicketItemContainer.displayName = "AddTicketItemContainer";
