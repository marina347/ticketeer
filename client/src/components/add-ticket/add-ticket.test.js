import React from "react";
import { shallow, mount } from "enzyme";

import { AddTicket } from "./add-ticket.component";

describe("AddTicket component", () => {
  let wrapper;
  let mockProps;
  let addTicket;
  let event;
  let currentUser;

  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, "useState");
  useStateSpy.mockImplementation((init) => [init, setState]);

  beforeEach(() => {
    currentUser = { id: "1" };
    addTicket = jest.fn();
    event = {
      preventDefault: jest.fn(),
      target: {
        name: "ticketName",
        value: "testName",
      },
    };
    mockProps = {
      addTicket,
      currentUser,
    };

    wrapper = shallow(<AddTicket {...mockProps} />);
  });

  it("Should render AddTicket component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("Should start onSubmit method when form inside is submitted", () => {
    //cant save duplicated lines to one variable and reuse it because change of prop is not reflected on that variable
    //like this const form =  wrapper.find("AddTicketItemContainer").find('[id="add_ticket_form"]') wont work below
    wrapper.find('[id="add_ticket_form"]').simulate("submit", event);
    expect(addTicket).toHaveBeenCalled();

    expect(
      wrapper
        .find('[id="add_ticket_form"]')
        .find('[id="add_ticket_input"]')
        .prop("value")
    ).toBe("");
  });

  it("Form text area has a prop onChange", () => {
    expect(
      wrapper
        .find('[id="add_ticket_form"]')
        .find('[id="add_ticket_input"]')
        .prop("onChange")
    ).not.toEqual(undefined);
  });
});
