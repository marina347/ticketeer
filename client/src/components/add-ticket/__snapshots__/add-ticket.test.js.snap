// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTicket component Should render AddTicket component 1`] = `
<AddTicket
  addTicket={[MockFunction]}
  currentUser={
    Object {
      "id": "1",
    }
  }
>
  <form
    className="add-ticket"
    id="add_ticket_form"
    onSubmit={[Function]}
  >
    <FormTextArea
      cols={50}
      id="add_ticket_input"
      label="Ticket name"
      name="ticketName"
      onChange={[Function]}
      placeholder="Ticket name"
      required={true}
      rows={3}
      value=""
    >
      <div
        class="form-input"
      >
        <textarea
          className="form-input__input"
          cols={50}
          id="add_ticket_input"
          name="ticketName"
          onChange={[Function]}
          placeholder="Ticket name"
          required={true}
          rows={3}
          value=""
        />
        <label
          className="form-input__label"
          id="form-textarea-label"
        >
          Ticket name
        </label>
      </div>
    </FormTextArea>
    <FormButton
      className="btn btn btn-main btn-main--white btn-main--small u-animation-none"
      type="submit"
    >
      <button
        className="btn btn btn-main btn-main--white btn-main--small u-animation-none"
        type="submit"
      >
        <span>
          ADD
        </span>
      </button>
    </FormButton>
  </form>
</AddTicket>
`;
